In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
patients = pd.read_csv('PATIENTS.csv')

In [3]:
patients.sample(15)

ROW_ID  SUBJECT_ID GENDER                  DOB                  DOD  \
27977   40436       76646      F  2072-11-08 00:00:00                  NaN   
15709   19215       20365      F  2091-06-21 00:00:00  2166-07-15 00:00:00   
39275   41153       79437      F  2089-08-15 00:00:00                  NaN   
11534   10137       10717      M  2147-12-14 00:00:00                  NaN   
21311   19572       20742      M  2052-03-07 00:00:00  2136-08-03 00:00:00   
27403   16782       17750      M  2059-10-04 00:00:00  2102-12-13 00:00:00   
15492   16463       17403      M  2157-02-02 00:00:00                  NaN   
46401   40717       77691      F  2100-06-03 00:00:00  2176-02-06 00:00:00   
39719   46399       99521      F  2126-07-12 00:00:00                  NaN   
39992   22969       24337      M  2040-05-31 00:00:00                  NaN   
34257   37047       63607      M  2062-06-29 00:00:00                  NaN   
6699     5915        6247      F  2188-04-28 00:00:00                  NaN   
3871     2843        3004      F  2160-04-13 00:00:00                  NaN   
201       807         857      M  2149-12-15 00:00:00                  NaN   
41413   41573       81063      M  2046-08-07 00:00:00                  NaN   

                  DOD_HOSP              DOD_SSN  EXPIRE_FLAG  
27977                  NaN                  NaN            0  
15709  2166-07-15 00:00:00  2166-07-15 00:00:00            1  
39275                  NaN                  NaN            0  
11534                  NaN                  NaN            0  
21311  2136-08-03 00:00:00  2136-08-03 00:00:00            1  
27403  2102-12-13 00:00:00                  NaN            1  
15492                  NaN                  NaN            0  
46401  2176-02-06 00:00:00  2176-02-06 00:00:00            1  
39719                  NaN                  NaN            0  
39992                  NaN                  NaN            0  
34257                  NaN                  NaN            0  
6699                   NaN                  NaN            0  
3871                   NaN                  NaN            0  
201                    NaN                  NaN            0  
41413                  NaN                  NaN            0

In [4]:
patients.dtypes

ROW_ID          int64
SUBJECT_ID      int64
GENDER         object
DOB            object
DOD            object
DOD_HOSP       object
DOD_SSN        object
EXPIRE_FLAG     int64
dtype: object

In [5]:
patients.EXPIRE_FLAG.value_counts()

0    30761
1    15759
Name: EXPIRE_FLAG, dtype: int64

In [6]:
patients.describe()

ROW_ID    SUBJECT_ID   EXPIRE_FLAG
count  46520.000000  46520.000000  46520.000000
mean   23260.500000  34425.772872      0.338758
std    13429.311598  28330.400343      0.473292
min        1.000000      2.000000      0.000000
25%    11630.750000  12286.750000      0.000000
50%    23260.500000  24650.500000      0.000000
75%    34890.250000  55477.500000      1.000000
max    46520.000000  99999.000000      1.000000

In [7]:
adm = pd.read_csv('ADMISSIONS.csv')

In [8]:
adm.sample(5)

ROW_ID  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME  \
24587   21597       17674   161979  2104-10-09 20:36:00  2104-10-16 13:57:00   
14526   17465       14259   184115  2109-10-14 20:02:00  2109-11-05 16:50:00   
26892   25929       21243   118125  2200-09-05 05:59:00  2200-09-07 14:33:00   
5052     5926        4894   177998  2130-06-06 21:02:00  2130-08-18 13:14:00   
16003   13522       11046   124034  2194-11-22 08:53:00  2194-12-19 16:00:00   

                 DEATHTIME ADMISSION_TYPE         ADMISSION_LOCATION  \
24587                  NaN      EMERGENCY       EMERGENCY ROOM ADMIT   
14526                  NaN      EMERGENCY  TRANSFER FROM HOSP/EXTRAM   
26892                  NaN        NEWBORN  PHYS REFERRAL/NORMAL DELI   
5052   2130-08-18 13:14:00      EMERGENCY       EMERGENCY ROOM ADMIT   
16003                  NaN        NEWBORN  CLINIC REFERRAL/PREMATURE   

              DISCHARGE_LOCATION INSURANCE LANGUAGE      RELIGION  \
24587  DISC-TRAN CANCER/CHLDRN H   Private      NaN        JEWISH   
14526   REHAB/DISTINCT PART HOSP  Medicare      NaN      CATHOLIC   
26892                       HOME   Private      NaN  UNOBTAINABLE   
5052                DEAD/EXPIRED   Private      NaN      CATHOLIC   
16003        SHORT TERM HOSPITAL   Private      NaN  UNOBTAINABLE   

      MARITAL_STATUS              ETHNICITY            EDREGTIME  \
24587        MARRIED  UNKNOWN/NOT SPECIFIED  2104-10-09 14:38:00   
14526        MARRIED                  WHITE                  NaN   
26892            NaN                  OTHER                  NaN   
5052         MARRIED                  WHITE  2130-06-06 13:35:00   
16003            NaN                  WHITE                  NaN   

                 EDOUTTIME              DIAGNOSIS  HOSPITAL_EXPIRE_FLAG  \
24587  2104-10-09 20:18:00  ALTERED MENTAL STATUS                     0   
14526                  NaN                EMPYEMA                     0   
26892                  NaN                NEWBORN                     0   
5052   2130-06-06 21:21:00      SUBDURAL HEMATOMA                     1   
16003                  NaN                NEWBORN                     0   

       HAS_CHARTEVENTS_DATA  
24587                     1  
14526                     1  
26892                     1  
5052                      1  
16003                     1

In [9]:
diag = pd.read_csv('DIAGNOSES_ICD.csv')

In [10]:
diag.sample(5)

ROW_ID  SUBJECT_ID  HADM_ID  SEQ_NUM ICD9_CODE
508398  498063       65900   175958     12.0     27652
147563  175532       15770   184670      5.0      2859
494620  510154       68490   147872     14.0     28529
166830  163891       14667   172924      2.0      4280
448314  444124       54197   154990      2.0      4275

In [12]:
diag.ICD9_CODE.value_counts().head(10)

4019     20703
4280     13111
42731    12891
41401    12429
5849      9119
25000     9058
2724      8690
51881     7497
5990      6555
53081     6326
Name: ICD9_CODE, dtype: int64

In [13]:
adm.DIAGNOSIS.value_counts().head(10)
#NOTE: these are hospital admission "diagnoses," NOT fully diagnosed conditions. Read: "purpose of this one hospital visit."

NEWBORN                                                      7823
PNEUMONIA                                                    1566
SEPSIS                                                       1184
CONGESTIVE HEART FAILURE                                      928
CORONARY ARTERY DISEASE                                       840
CHEST PAIN                                                    778
INTRACRANIAL HEMORRHAGE                                       713
ALTERED MENTAL STATUS                                         712
GASTROINTESTINAL BLEED                                        686
CORONARY ARTERY DISEASE\CORONARY ARTERY BYPASS GRAFT /SDA     583
Name: DIAGNOSIS, dtype: int64

In [14]:
diagdict = pd.read_csv('D_ICD_DIAGNOSES.csv')

In [15]:
diagdict.sample(10)

ROW_ID ICD9_CODE               SHORT_TITLE  \
3190     3191     28866                  Bandemia   
4786     4785     37641    Hypertelorism of orbit   
10342   11441     99665  React-oth genitourin dev   
9046     9047      7715   Neonatal infec mastitis   
4928     4927     37943  Mydriasis not d/t mydrtc   
1636     1639      2631  Malnutrition mild degree   
14441   14311     94555    3 deg burn w loss-knee   
6172     6176      5523  Diaphragm hernia w obstr   
8880     8881     71909   Joint effusion-mult jts   
13381   13954     V1270  Prsnl hst unspc dgstv ds   

                                              LONG_TITLE  
3190                                            Bandemia  
4786                              Hypertelorism of orbit  
10342  Infection and inflammatory reaction due to oth...  
9046                         Neonatal infective mastitis  
4928       Mydriasis (persistent), not due to mydriatics  
1636                         Malnutrition of mild degree  
14441  Deep necrosis of underlying tissues [deep thir...  
6172               Diaphragmatic hernia with obstruction  
8880                   Effusion of joint, multiple sites  
13381  Personal history of unspecified digestive disease

In [16]:
diagnosis = diag.merge(diagdict, on = 'ICD9_CODE', how = 'left')

In [18]:
diagnosis.LONG_TITLE.value_counts().head(10)
#NOTE: THESE are the top 10 diagnoses for patients in this hospital. 

Unspecified essential hypertension                                                                            20703
Congestive heart failure, unspecified                                                                         13111
Atrial fibrillation                                                                                           12891
Coronary atherosclerosis of native coronary artery                                                            12429
Acute kidney failure, unspecified                                                                              9119
Diabetes mellitus without mention of complication, type II or unspecified type, not stated as uncontrolled     9058
Other and unspecified hyperlipidemia                                                                           8690
Acute respiratory failure                                                                                      7497
Urinary tract infection, site not specified                             

In [19]:
diagnosis.ICD9_CODE.value_counts().head(10)
#...and their corresponding ICD9_CODEs

4019     20703
4280     13111
42731    12891
41401    12429
5849      9119
25000     9058
2724      8690
51881     7497
5990      6555
53081     6326
Name: ICD9_CODE, dtype: int64

In [22]:
diagnosis[diagnosis.iloc[:,4] == '4280'].SUBJECT_ID.nunique()
#We will go with Congestive Heart Failure as it is the only non-curable condition in the top 3. 

9843

In [23]:
9843/46520

0.21158641444539983

Note: 9,843 patients (21.2%) out of 46,520 in this hospital database have Congestive Heart Failure. This is our baseline accuracy.

In [24]:
#CAUTION: SOME PATIENTS HAVE NUMEROUS DISEASES
diagnosis.head(28)

ROW_ID_x  SUBJECT_ID  HADM_ID  SEQ_NUM ICD9_CODE  ROW_ID_y  \
0       1297         109   172335      1.0     40301    4312.0   
1       1298         109   172335      2.0       486    5528.0   
2       1299         109   172335      3.0     58281    5892.0   
3       1300         109   172335      4.0      5855    5913.0   
4       1301         109   172335      5.0      4254    4435.0   
5       1302         109   172335      6.0      2762    2787.0   
6       1303         109   172335      7.0      7100    8151.0   
7       1304         109   172335      8.0      2767    2159.0   
8       1305         109   172335      9.0      7243   12715.0   
9       1306         109   172335     10.0     45829    5394.0   
10      1307         109   172335     11.0      2875    3169.0   
11      1308         109   172335     12.0     28521    3142.0   
12      1309         109   172335     13.0     28529    3144.0   
13      1310         109   172335     14.0     27541    2779.0   
14      1311         109   173633      1.0     40301    4312.0   
15      1312         109   173633      2.0      5856    5914.0   
16      1313         109   173633      3.0     58381    5901.0   
17      1314         109   173633      4.0      7100    8151.0   
18      1315         109   173633      5.0      5589    5582.0   
19      1316         109   173633      6.0      2875    3169.0   
20      1317         109   173633      7.0     28521    3142.0   
21      1318         109   173633      8.0     45829    5394.0   
22      1319         109   173633      9.0     32723    3119.0   
23      1320         109   173633     10.0     22804    2521.0   
24      1321         109   173633     11.0     33829    3414.0   
25      1322         109   173633     12.0     78900   13887.0   
26      1323         109   173633     13.0     79092   10026.0   
27      1324         109   173633     14.0     V4511    9885.0   

                 SHORT_TITLE  \
0     Mal hyp kid w cr kid V   
1    Pneumonia, organism NOS   
2   Chr nephritis in oth dis   
3   Chron kidney dis stage V   
4    Prim cardiomyopathy NEC   
5                   Acidosis   
6   Syst lupus erythematosus   
7            Hyperpotassemia   
8                   Sciatica   
9   Iatrogenc hypotnsion NEC   
10      Thrombocytopenia NOS   
11  Anemia in chr kidney dis   
12  Anemia-other chronic dis   
13              Hypocalcemia   
14    Mal hyp kid w cr kid V   
15   End stage renal disease   
16  Nephritis NOS in oth dis   
17  Syst lupus erythematosus   
18  Noninf gastroenterit NEC   
19      Thrombocytopenia NOS   
20  Anemia in chr kidney dis   
21  Iatrogenc hypotnsion NEC   
22   Obstructive sleep apnea   
23    Hemangioma intra-abdom   
24          Chronic pain NEC   
25  Abdmnal pain unspcf site   
26  Abnrml coagultion prfile   
27     Renal dialysis status   

                                           LONG_TITLE  
0   Hypertensive chronic kidney disease, malignant...  
1                     Pneumonia, organism unspecified  
2   Chronic glomerulonephritis in diseases classif...  
3                     Chronic kidney disease, Stage V  
4                      Other primary cardiomyopathies  
5                                            Acidosis  
6                        Systemic lupus erythematosus  
7                                     Hyperpotassemia  
8                                            Sciatica  
9                        Other iatrogenic hypotension  
10                      Thrombocytopenia, unspecified  
11                   Anemia in chronic kidney disease  
12                    Anemia of other chronic disease  
13                                       Hypocalcemia  
14  Hypertensive chronic kidney disease, malignant...  
15                            End stage renal disease  
16  Nephritis and nephropathy, not specified as ac...  
17                       Systemic lupus erythematosus  
18  Other and unspecified noninfectious gastroente...  
19                      Thrombocytopenia

## CAUTION: HEAVY CELLS AHEAD

In [25]:
#chunk_list = list(pd.read_csv('LABEVENTS.csv',chunksize=10e6))
#df0 = chunk_list[0]
#df1 = chunk_list[1]
#df2 = chunk_list[2]
labevents = pd.read_csv('LABEVENTS.csv')

In [26]:
labevents.head()

ROW_ID  SUBJECT_ID  HADM_ID  ITEMID            CHARTTIME VALUE  VALUENUM  \
0     281           3      NaN   50820  2101-10-12 16:07:00  7.39      7.39   
1     282           3      NaN   50800  2101-10-12 18:17:00   ART       NaN   
2     283           3      NaN   50802  2101-10-12 18:17:00    -1     -1.00   
3     284           3      NaN   50804  2101-10-12 18:17:00    22     22.00   
4     285           3      NaN   50808  2101-10-12 18:17:00  0.93      0.93   

  VALUEUOM      FLAG  
0    units       NaN  
1      NaN       NaN  
2    mEq/L       NaN  
3    mEq/L       NaN  
4   mmol/L  abnormal

In [27]:
labevents.shape
#27 million rows...yikes!

(27854055, 9)

In [110]:
#CAUTION: THIS CELL IS A TEST; CELL TAKES 5 MIN TO RUN
'''
chunksize = 10e6
counter = 0
for chunk in pd.read_csv('LABEVENTS.csv', chunksize = chunksize):
    if counter ==0:
        df1 = hunk
    elif counter ==1:
        df2 = chunk
    
    #print(chunk.head(1))
    '''

   ROW_ID  SUBJECT_ID  HADM_ID  ITEMID            CHARTTIME VALUE  VALUENUM  \
0     281           3      NaN   50820  2101-10-12 16:07:00  7.39      7.39   

  VALUEUOM FLAG  
0    units  NaN  
            ROW_ID  SUBJECT_ID   HADM_ID  ITEMID            CHARTTIME VALUE  \
10000000  10149477       16410  168812.0   50970  2140-12-25 02:32:00   2.8   

          VALUENUM VALUEUOM FLAG  
10000000       2.8    mg/dL  NaN  
            ROW_ID  SUBJECT_ID   HADM_ID  ITEMID            CHARTTIME VALUE  \
20000000  19935592       40175  103117.0   50878  2189-10-12 07:30:00    67   

          VALUENUM VALUEUOM      FLAG  
20000000      67.0     IU/L  abnormal  


Before any serious aggregation, let's see what measurements (labevents) are taken the most.

Side objective: we may try to reduce the size of the large labevents DF by filtering out low frequency measurements.

In [30]:
D_LABITEMS = pd.read_csv('D_LABITEMS.csv')

In [32]:
D_LABITEMS.head()

ROW_ID  ITEMID                       LABEL                      FLUID  \
0     546   51346                      Blasts  Cerebrospinal Fluid (CSF)   
1     547   51347                 Eosinophils  Cerebrospinal Fluid (CSF)   
2     548   51348             Hematocrit, CSF  Cerebrospinal Fluid (CSF)   
3     549   51349  Hypersegmented Neutrophils  Cerebrospinal Fluid (CSF)   
4     550   51350           Immunophenotyping  Cerebrospinal Fluid (CSF)   

     CATEGORY LOINC_CODE  
0  Hematology    26447-3  
1  Hematology    26451-5  
2  Hematology    30398-2  
3  Hematology    26506-6  
4  Hematology        NaN

In [40]:
D_LABITEMS.dtypes

ROW_ID         int64
ITEMID         int64
LABEL         object
FLUID         object
CATEGORY      object
LOINC_CODE    object
dtype: object

In [33]:
lab_measurements_test = labevents.merge(D_LABITEMS, left_index = True, on = "ITEMID", how = 'left')

In [35]:
lab_measurements_test.ITEMID.value_counts().head(35)
#NOTE: big dropoff after 400k value counts (top 29 measurements)

51221    881764
50971    845737
50983    808401
50912    797389
50902    795480
51006    791838
50882    780648
51265    778365
50868    769810
51301    753221
51222    752444
50931    748896
51249    748147
51279    747999
51248    747994
51250    747977
51277    746817
50960    664123
50893    591932
50970    590502
50820    530752
50802    490651
50804    490641
50821    490628
50818    490594
51275    474930
51237    471176
51274    469083
50800    404785
50808    249110
50885    238263
50861    219462
50878    219452
50863    207847
50809    196736
Name: ITEMID, dtype: int64

In [42]:
ITEMID_list = [51221,50971,50983,50912,50902,51006,50882,51265,50868,51301,51222,50931,51249,51279,\
51248,51250,51277,50960,50893,50970,50820,50802,50804,50821,50818,51275,51237,51274,50800]

In [49]:
reduced_d_labitems = D_LABITEMS[D_LABITEMS.ITEMID.isin(ITEMID_list)]
reduced_d_labitems
#LIST OF 29 MOST COMMON LAB TESTS ON ALL PATIENTS

ROW_ID  ITEMID                 LABEL  FLUID    CATEGORY LOINC_CODE
127       1   50800         SPECIMEN TYPE  BLOOD   BLOOD GAS        NaN
129       3   50802           Base Excess  Blood   Blood Gas    11555-0
131       5   50804  Calculated Total CO2  Blood   Blood Gas    34728-6
145      19   50818                  pCO2  Blood   Blood Gas    11557-6
147      21   50820                    pH  Blood   Blood Gas    11558-4
148      22   50821                   pO2  Blood   Blood Gas    11556-8
195      69   50868             Anion Gap  Blood   Chemistry     1863-0
209      83   50882           Bicarbonate  Blood   Chemistry     1963-8
220      94   50893        Calcium, Total  Blood   Chemistry     2000-8
229     103   50902              Chloride  Blood   Chemistry     2075-0
239     113   50912            Creatinine  Blood   Chemistry     2160-0
258     132   50931               Glucose  Blood   Chemistry     2345-7
287     161   50960             Magnesium  Blood   Chemistry     2601-3
297     171   50970             Phosphate  Blood   Chemistry     2777-1
298     172   50971             Potassium  Blood   Chemistry     2823-3
310     184   50983                Sodium  Blood   Chemistry     2951-2
332     206   51006         Urea Nitrogen  Blood   Chemistry     3094-0
547     421   51221            Hematocrit  Blood  Hematology     4544-3
548     422   51222            Hemoglobin  Blood  Hematology      718-7
563     437   51237               INR(PT)  Blood  Hematology     5895-7
574     448   51248                   MCH  Blood  Hematology      785-6
575     449   51249                  MCHC  Blood  Hematology      786-4
576     450   51250                   MCV  Blood  Hematology      787-2
591     465   51265        Platelet Count  Blood  Hematology      777-3
600     474   51274                    PT  Blood  Hematology     5902-2
601     475   51275                   PTT  Blood  Hematology     3173-2
603     477   51277                   RDW  Blood  Hematology      788-0
605     479   51279       Red Blood Cells  Blood  Hematology      789-8
627     501   51301     White Blood Cells  Blood  Hematology      804-5

In [50]:
reduced_labevents = labevents.merge(reduced_d_labitems, left_index = True, on = "ITEMID", how = 'inner')

In [51]:
reduced_labevents.shape
#Not much smaller :( womp womp

(19402724, 14)

In [52]:
reduced_labevents.head()

ROW_ID_x  SUBJECT_ID   HADM_ID  ITEMID            CHARTTIME VALUE  \
147       281           3       NaN   50820  2101-10-12 16:07:00  7.39   
147       289           3       NaN   50820  2101-10-12 18:17:00  7.42   
147       396           3       NaN   50820  2101-10-15 12:42:00  7.49   
147       976           3  145834.0   50820  2101-10-22 04:31:00  7.40   
147       989           3  145834.0   50820  2101-10-22 07:13:00  7.37   

     VALUENUM VALUEUOM FLAG  ROW_ID_y LABEL  FLUID   CATEGORY LOINC_CODE  
147      7.39    units  NaN        21    pH  Blood  Blood Gas    11558-4  
147      7.42    units  NaN        21    pH  Blood  Blood Gas    11558-4  
147      7.49    units  NaN        21    pH  Blood  Blood Gas    11558-4  
147      7.40    units  NaN        21    pH  Blood  Blood Gas    11558-4  
147      7.37    units  NaN        21    pH  Blood  Blood Gas    11558-4

In [101]:
#CAUTION: CELL TAKES 20 MIN+ TO RUN
#for chunk in pd.read_csv('LABEVENTS.csv', chunksize = chunksize):
#    lab_group = chunk.groupby(['SUBJECT_ID', 'ITEMID', 'VALUEUOM']).agg(['min', 'mean', 'max'])['VALUENUM']
lab_group = reduced_labevents.groupby(['SUBJECT_ID', 'ITEMID']).agg(['min', 'mean', 'max', 'std', 'var'])['VALUENUM'].reset_index()

In [102]:
lab_group.head(30)

SUBJECT_ID  ITEMID     min        mean     max         std           var
0            2   51221    0.00   24.800000   49.60   35.072496   1230.080000
1            2   51222    0.00    8.250000   16.50   11.667262    136.125000
2            2   51248    0.00   17.200000   34.40   24.324473    591.680000
3            2   51249    0.00   16.650000   33.30   23.546656    554.445000
4            2   51250    0.00   51.500000  103.00   72.831998   5304.500000
5            2   51265    5.00  153.500000  302.00  210.010714  44104.500000
6            2   51277    0.00    7.750000   15.50   10.960155    120.125000
7            2   51279    0.00    2.405000    4.81    3.401184     11.568050
8            2   51301    0.10   11.050000   22.00   15.485639    239.805000
9            3   50800     NaN         NaN     NaN         NaN           NaN
10           3   50802  -12.00   -4.266667    6.00    3.774315     14.245455
11           3   50804   13.00   19.844444   34.00    4.237042     17.952525
12           3   50818   23.00   31.600000   50.00    5.690502     32.381818
13           3   50820    7.26    7.388958    7.59    0.054509      0.002971
14           3   50821   20.00  114.955556  329.00   76.154137   5799.452525
15           3   50868    9.00   14.705882   23.00    2.713931      7.365419
16           3   50882   11.00   22.085714   27.00    4.578540     20.963025
17           3   50893    5.80    7.925926    9.10    0.699226      0.488917
18           3   50902   96.00  105.638889  112.00    3.810907     14.523016
19           3   50912    1.10    1.647368    3.20    0.424733      0.180398
20           3   50931   73.00  121.714286  293.00   47.827860   2287.504202
21           3   50960    1.60    2.063636    3.50    0.350730      0.123011
22           3   50970    2.50    3.485714    5.60    0.812274      0.659788
23           3   50971    3.40    4.316667    7.30    0.651421      0.424350
24           3   50983  125.00  138.027027  143.00    3.303924     10.915916
25           3   51006   15.00   33.236842   56.00   11.090376    122.996444
26           3   51221   24.90   30.818182   38.40    2.996399      8.978409
27           3   51222    7.80   10.278125   12.80    1.060047      1.123700
28           3   51237    1.10    1.407692    1.70    0.201914      0.040769
29           3   51248   28.80   29.643750   30.80    0.551208      0.303831

In [103]:
lab_group.nunique()

SUBJECT_ID     45541
ITEMID            29
min             2173
mean          276384
max             3465
std           650619
var           691436
dtype: int64

In [104]:
lab_group2 = lab_group.merge(reduced_d_labitems, left_index = True, on = "ITEMID", how = 'left')
lab_group2.head()

SUBJECT_ID  ITEMID  min   mean    max        std       var  ROW_ID  \
547           2   51221  0.0  24.80   49.6  35.072496  1230.080     421   
548           2   51222  0.0   8.25   16.5  11.667262   136.125     422   
574           2   51248  0.0  17.20   34.4  24.324473   591.680     448   
575           2   51249  0.0  16.65   33.3  23.546656   554.445     449   
576           2   51250  0.0  51.50  103.0  72.831998  5304.500     450   

          LABEL  FLUID    CATEGORY LOINC_CODE  
547  Hematocrit  Blood  Hematology     4544-3  
548  Hemoglobin  Blood  Hematology      718-7  
574         MCH  Blood  Hematology      785-6  
575        MCHC  Blood  Hematology      786-4  
576         MCV  Blood  Hematology      787-2

In [100]:
#test[test['SUBJECT_ID']==3]

In [118]:
#test = lab_group2[lab_group2.SUBJECT_ID.isin(lab_group2.SUBJECT_ID.unique())][['SUBJECT_ID', 'LABEL', 'min', 'mean', 'max', 'std', 'var']]
#test.pivot(index='SUBJECT_ID', columns='LABEL')

In [105]:
#labgroup_pivot = lab_group2.pivot(index='SUBJECT_ID', columns='LABEL')
labgroup_pivot = lab_group2.pivot(index='SUBJECT_ID', columns='LABEL', values=['min', 'mean', 'max', 'std', 'var'])

In [106]:
labgroup_pivot.head()

min                                         \
LABEL      Anion Gap Base Excess Bicarbonate Calcium, Total   
SUBJECT_ID                                                    
2                NaN         NaN         NaN            NaN   
3                9.0       -12.0        11.0            5.8   
4               13.0        -5.0        16.0            7.7   
5                NaN         NaN         NaN            NaN   
6               10.0        -9.0        14.0            6.8   

                                                                        \
LABEL      Calculated Total CO2 Chloride Creatinine Glucose Hematocrit   
SUBJECT_ID                                                               
2                           NaN      NaN        NaN     NaN        0.0   
3                          13.0     96.0        1.1    73.0       24.9   
4                          18.0     86.0        0.4    60.0       19.6   
5                           NaN      NaN        NaN     NaN       43.0   
6                          16.0     95.0        0.7    95.0       22.4   

                       ...       var                              \
LABEL      Hemoglobin  ... Potassium         RDW Red Blood Cells   
SUBJECT_ID             ...                                         
2                 0.0  ...       NaN  120.125000       11.568050   
3                 7.8  ...  0.424350    0.663216        0.115032   
4                 6.8  ...  0.193825    3.160421        0.128649   
5                14.9  ...       NaN         NaN             NaN   
6                 7.0  ...  0.316837    1.245496        0.187785   

                                                                     \
LABEL      SPECIMEN TYPE     Sodium Urea Nitrogen White Blood Cells   
SUBJECT_ID                                                            
2                    NaN        NaN           NaN        239.805000   
3                    NaN  10.915916    122.996444         16.243942   
4                    NaN  13.936170     32.573295         11.198652   
5                    NaN        NaN           NaN               NaN   
6                    NaN  16.861895    719.225806         41.266101   

                                              
LABEL            pCO2        pH          pO2  
SUBJECT_ID                                    
2                 NaN       NaN          NaN  
3           32.381818  0.002971  5799.452525  
4            0.500000  0.006533   512.000000  
5                 NaN       NaN          NaN  
6           27.700000  0.004120   432.300000  

[5 rows x 145 columns]

In [107]:
labgroup_pivot.columns = list(map(lambda x: x[1]+x[0], zip(labgroup_pivot.columns.get_level_values(0), map(str, labgroup_pivot.columns.get_level_values(1)))))

In [108]:
labgroup_pivot.head(10)

Anion Gapmin  Base Excessmin  Bicarbonatemin  Calcium, Totalmin  \
SUBJECT_ID                                                                    
2                    NaN             NaN             NaN                NaN   
3                    9.0           -12.0            11.0                5.8   
4                   13.0            -5.0            16.0                7.7   
5                    NaN             NaN             NaN                NaN   
6                   10.0            -9.0            14.0                6.8   
7                    NaN             NaN             NaN                NaN   
8                    NaN             NaN             NaN                NaN   
9                   10.0            -1.0            21.0                7.8   
10                  17.0             NaN            18.0                NaN   
11                  12.0             2.0            25.0                8.8   

            Calculated Total CO2min  Chloridemin  Creatininemin  Glucosemin  \
SUBJECT_ID                                                                    
2                               NaN          NaN            NaN         NaN   
3                              13.0         96.0            1.1        73.0   
4                              18.0         86.0            0.4        60.0   
5                               NaN          NaN            NaN         NaN   
6                              16.0         95.0            0.7        95.0   
7                               NaN          NaN            NaN         NaN   
8                               NaN          NaN            NaN         NaN   
9                              22.0         98.0            1.2       114.0   
10                              NaN         98.0            NaN         NaN   
11                             25.0         97.0            0.5        92.0   

            Hematocritmin  Hemoglobinmin  ...  Potassiumvar      RDWvar  \
SUBJECT_ID                                ...                             
2                     0.0            0.0  ...           NaN  120.125000   
3                    24.9            7.8  ...      0.424350    0.663216   
4                    19.6            6.8  ...      0.193825    3.160421   
5                    43.0           14.9  ...           NaN         NaN   
6                    22.4            7.0  ...      0.316837    1.245496   
7                    53.0           17.6  ...           NaN         NaN   
8                    52.0           17.5  ...           NaN         NaN   
9                    33.5           11.5  ...      0.123222    0.177143   
10                   42.8           14.8  ...      0.045000         NaN   
11                   33.0           11.3  ...      0.099943    0.744506   

            Red Blood Cellsvar  SPECIMEN TYPEvar  Sodiumvar  Urea Nitrogenvar  \
SUBJECT_ID                                                                      
2                    11.568050               NaN        NaN               NaN   
3                     0.115032               NaN  10.915916        122.996444   
4                     0.128649               NaN  13.936170         32.573295   
5                          NaN               NaN        NaN               NaN   
6                     0.187785               NaN  16.861895        719.225806   
7                          NaN               NaN        NaN               NaN   
8                          NaN               NaN        NaN               NaN   
9                     0.186855               NaN  12.455556         36.571429   
10                         NaN               NaN  50.000000               NaN   
11                    0.072453               NaN   4.174603          9.011396   

            White Blood Cellsvar     pCO2var     pHvar       pO2var  
SUBJECT_ID                                                           
2                     239.805000         NaN       NaN          NaN  
3                      16.2439

In [117]:
labgroup_pivot.isnull().sum()

Anion Gapmin                4372
Base Excessmin             14114
Bicarbonatemin              4361
Calcium, Totalmin           7548
Calculated Total CO2min    14114
Chloridemin                 4338
Creatininemin               6192
Glucosemin                  6743
Hematocritmin                158
Hemoglobinmin                244
INR(PT)min                  7748
MCHmin                       252
MCHCmin                      250
MCVmin                       252
Magnesiummin                7156
PTmin                       7749
PTTmin                      7843
Phosphatemin                7629
Platelet Countmin            238
Potassiummin                4333
RDWmin                       258
Red Blood Cellsmin           252
SPECIMEN TYPEmin           45541
Sodiummin                   4332
Urea Nitrogenmin            6180
White Blood Cellsmin         207
pCO2min                    14117
pHmin                      13064
pO2min                     14117
Anion Gapmean               4372
          

In [122]:
patients_labevents = patients.join(labgroup_pivot, how = 'left', on = 'SUBJECT_ID')

In [150]:
patients_labevents.head()

ROW_ID  SUBJECT_ID GENDER                  DOB                  DOD  \
0     234         249      F  2075-03-13 00:00:00                  NaN   
1     235         250      F  2164-12-27 00:00:00  2188-11-22 00:00:00   
2     236         251      M  2090-03-15 00:00:00                  NaN   
3     237         252      M  2078-03-06 00:00:00                  NaN   
4     238         253      F  2089-11-26 00:00:00                  NaN   

              DOD_HOSP DOD_SSN  EXPIRE_FLAG  Anion Gapmin  Base Excessmin  \
0                  NaN     NaN            0           8.0            -1.0   
1  2188-11-22 00:00:00     NaN            1           8.0           -23.0   
2                  NaN     NaN            0          14.0             NaN   
3                  NaN     NaN            0           6.0            -8.0   
4                  NaN     NaN            0          10.0             1.0   

   ...  Potassiumvar    RDWvar  Red Blood Cellsvar  SPECIMEN TYPEvar  \
0  ...      0.108914  0.915249            0.126727               NaN   
1  ...      0.900686  2.933919            0.136234               NaN   
2  ...      0.070000  0.043333            0.019033               NaN   
3  ...      0.426788  2.200418            0.221759               NaN   
4  ...      0.040000  0.023000            0.015320               NaN   

   Sodiumvar  Urea Nitrogenvar  White Blood Cellsvar     pCO2var     pHvar  \
0   8.221154        133.147541             28.458241   98.423333  0.006568   
1   9.871658          9.134581             50.243707  122.549142  0.014248   
2   4.333333          1.666667              9.143333         NaN       NaN   
3  13.572590         61.720930             22.781957   11.339181  0.006636   
4   4.566667         14.266667              4.943000    8.916667  0.000892   

         pO2var  
0   9958.000000  
1  17682.562724  
2           NaN  
3   2751.397661  
4  22969.000000  

[5 rows x 153 columns]

In [123]:
#checkpoint export
patients_labevents.to_csv('patients_labevents.csv')

Now to join in target to Patient Dataset.

In [126]:
CHF_group = diagnosis[diagnosis.iloc[:,4] == '4280']

In [137]:
CHF_group['SUBJECT_ID'].unique()

array([  115,   117,   124, ..., 97144, 97172, 97488], dtype=int64)

In [139]:
chf_df = pd.DataFrame()

In [141]:
chf_df['SUBJECT_ID'] = CHF_group['SUBJECT_ID'].unique()

In [143]:
chf_df['CHF'] = 1

In [148]:
chf_df.head()

SUBJECT_ID  CHF
0         115    1
1         117    1
2         124    1
3         130    1
4          68    1

In [154]:
chf_df.shape

(9843, 2)

In [157]:
patients_CHF = patients_labevents.merge(chf_df, how = 'left', on = 'SUBJECT_ID')
patients_CHF.shape

(46520, 154)

In [160]:
patients_CHF.CHF.isnull().sum()

36677

In [162]:
patients_CHF.CHF.fillna(0, inplace = True)

In [163]:
patients_CHF.CHF.value_counts()

0.0    36677
1.0     9843
Name: CHF, dtype: int64

In [164]:
patients_CHF.to_csv('patients_labevents.csv')

In [165]:
#CAUTION: CELL TAKES 20 MIN+ TO RUN
'''
for chunk in pd.read_csv('CHARTEVENTS.csv', chunksize = chunksize):
    chart_group = chunk.groupby(['SUBJECT_ID', 'ITEMID', 'VALUEUOM']).agg(['min', 'mean', 'max'])['VALUE']
'''

"\nfor chunk in pd.read_csv('CHARTEVENTS.csv', chunksize = chunksize):\n    chart_group = chunk.groupby(['SUBJECT_ID', 'ITEMID', 'VALUEUOM']).agg(['min', 'mean', 'max'])['VALUE']\n"

In [166]:
#chart_group.head()